modules/Bio/EnsEMBL/Compara/RunnableDB/OrthologQM/GeneOrderConservation.pm

```diff
@@ -0,0 +1,386 @@
+    return \@uniq_ids;
+}
+
+sub _calculate_goc {
+    my ( $self, $gm_id_a, $gm_id_b, $hom_id ) = @_;
+    
+    # get neighbourhood for each gene_member
+    my $num_neighbours = $self->param_required('number_neighbours');
+    my @neighbours_a   = $self->_get_neighbours( $gm_id_a, $num_neighbours );
+    # we want to allow a certain number of gaps, so more neighbours should be fetched for gene B
+    my $allowed_gap    = $self->param_required('allowed_gap');
+    my $wanted_neighbours = $num_neighbours + ( $allowed_gap * ($num_neighbours - 1) );
+    my @neighbours_b   = $self->_get_neighbours( $gm_id_b, $wanted_neighbours );
+    
+    my $gene_member_strand = $self->param('gene_member_strand');
+    my $strand_mismatch;
+    unless ($gene_member_strand->{$gm_id_a} eq $gene_member_strand->{$gm_id_b}) {
+        @neighbours_b = reverse @neighbours_b;
+        $strand_mismatch = 1;
+    }
+    
+    # check for overlaps between left side neighbours' homology ids
+    my $shared_homology = 0;
+    my $prev_match_pos = -1;
+    for ( my $i = 0; $i < @neighbours_a; $i++ ) {
+        my $neighbour_id_a = $neighbours_a[$i];
+        for ( my $j = $prev_match_pos+1; $j < @neighbours_b; $j++ ) {
+            my $neighbour_id_b = $neighbours_b[$j];
+            my $share_homology = $self->_find_shared_homology($neighbour_id_a, $neighbour_id_b, $strand_mismatch);
+            if ( $share_homology ) {
+                $shared_homology++ unless ( $prev_match_pos >= 0 && abs($j - $prev_match_pos) > $allowed_gap + 1);
+                $prev_match_pos = $j;
+                last; # don't care about multiple matches, so let's move on here
+            }
+        }
+    }
+    my $perc_overlap = ($shared_homology / ($num_neighbours*2)) * 100;
+    return $perc_overlap;
+}
+
+sub _get_neighbours {
+    my ( $self, $gm_id, $num_neighbours) = @_;
+
+    my ($dnafrag_id, $pos) = $self->_get_dnafrag_position($gm_id);
+    my @these_neighbours;
+    foreach my $x ( 1..$num_neighbours ) {
+        my $gm_left = $self->_get_gene_member_by_dnafrag_pos($dnafrag_id, ($pos-$x));
+        unshift @these_neighbours, $gm_left if defined $gm_left;
+        
+        my $gm_right = $self->_get_gene_member_by_dnafrag_pos($dnafrag_id, ($pos+$x));
+        push @these_neighbours, $gm_right if defined $gm_right;
+    }
+    return @these_neighbours;
+}
+
+sub _get_dnafrag_position {
+    my ( $self, $gene_member_id ) = @_;
+    
+    my $neighbourhood = $self->param('neighbourhood')->{'member2pos'};
+    my ( $dnafrag_id, $pos ) = @{ $neighbourhood->{$gene_member_id} };
+    
+    return ( $dnafrag_id, $pos );
+}
+
+sub _get_gene_member_by_dnafrag_pos {
+    my ( $self, $dnafrag_id, $pos ) = @_;
+    
+    my $neighbourhood = $self->param('neighbourhood')->{'pos2member'};
+    my $gene_member_id = $neighbourhood->{$dnafrag_id}->{$pos};
+    
+    return $gene_member_id;
+}
+
+sub _find_shared_homology {
+    my ( $self, $gm_id_a, $gm_id_b, $strand_mismatch ) = @_;
```

Comment:
Instead of repeatedly _grep_-ing pairs of gene_member_ids in the file. Could you not load the file in a hash ?

Reply:
For some reason, I thought memory would be an issue, but it really isn't, so I've implemented this suggestion (so much faster!). Thanks.